File: sdks/plugin-wire/java/io/amesh/plugin/Wire.java
package io.amesh.plugin;

import java.time.Instant;
import java.util.List;
import java.util.Map;

/** AMESH isolated-plugin wire contracts. The normative schema is schemas/plugin-wire.schema.json. */
public final class Wire {
    public static final String VERSION = "amesh.plugin.wire/v1";
    public static final String HANDSHAKE = "amesh.handshake";
    public static final String DISCOVER = "amesh.discover";
    public static final String VALIDATE = "amesh.validate";
    public static final String INVOKE = "amesh.invoke";
    public static final String CANCEL = "amesh.cancel";
    public static final String SHUTDOWN = "amesh.shutdown";
    public static final List<String> REQUIRED_FEATURES = List.of(
            "schema-discovery", "validation", "execution", "cancellation",
            "heartbeats", "logs", "metrics", "artifacts");

    private Wire() {}

    public record WorkloadIdentity(String sessionId, String workloadToken, Instant expiresAt) {}

    public record CapabilityEnvelope(
            Map<String, String> capabilityTokens,
            Map<String, String> secrets,
            Map<String, String> files,
            List<String> allowedEgress,
            List<String> platformApis) {}

    public record PluginRequest(
            String protocolVersion,
            String plugin,
            String entryPoint,
            String operation,
            Map<String, Object> session,
            Map<String, Object> configuration,
            Map<String, Object> input,
            Map<String, Object> context) {}

    public record InvocationParams(
            String sessionId,
            String workloadToken,
            PluginRequest request,
            CapabilityEnvelope capabilities) {}

    public record EntryPoint(
            String name,
            String type,
            String resourceType,
            Map<String, Object> configurationSchema,
            Map<String, Object> outputSchema) {}

    public record DiscoveryResult(
            String sessionId, String workloadToken, List<EntryPoint> entryPoints) {}

    public record PluginResponse(
            String protocolVersion,
            String invocationId,
            Map<String, Object> output,
            List<Map<String, Object>> logs,
            List<Map<String, Object>> errors,
            Map<String, Object> checkpoint) {}

    @FunctionalInterface
    public interface PluginHandler {
        PluginResponse invoke(PluginRequest request, CapabilityEnvelope capabilities) throws Exception;
    }
}

File: sdks/plugin-wire/go/wire.go
// Package ameshplugin defines AMESH isolated-plugin wire contracts.
// The normative schema is schemas/plugin-wire.schema.json.
package ameshplugin

const WireVersion = "amesh.plugin.wire/v1"

const (
	MethodHandshake = "amesh.handshake"
	MethodDiscover  = "amesh.discover"
	MethodValidate  = "amesh.validate"
	MethodInvoke    = "amesh.invoke"
	MethodCancel    = "amesh.cancel"
	MethodShutdown  = "amesh.shutdown"
)

var RequiredFeatures = []string{
	"schema-discovery",
	"validation",
	"execution",
	"cancellation",
	"heartbeats",
	"logs",
	"metrics",
	"artifacts",
}

type WorkloadIdentity struct {
	SessionID     string `json:"sessionId"`
	WorkloadToken string `json:"workloadToken"`
	ExpiresAt     string `json:"expiresAt"`
}

type CapabilityEnvelope struct {
	CapabilityTokens map[string]string `json:"capabilityTokens"`
	Secrets          map[string]string `json:"secrets"`
	Files            map[string]string `json:"files"`
	AllowedEgress    []string          `json:"allowedEgress"`
	PlatformAPIs     []string          `json:"platformApis"`
}

type PluginRequest struct {
	ProtocolVersion string                 `json:"protocolVersion"`
	Plugin          string                 `json:"plugin"`
	EntryPoint      string                 `json:"entryPoint"`
	Operation       string                 `json:"operation"`
	Session         map[string]any         `json:"session"`
	Configuration   map[string]any         `json:"configuration"`
	Input           map[string]any         `json:"input"`
	Context         map[string]any         `json:"context"`
}

type InvocationParams struct {
	SessionID     string             `json:"sessionId"`
	WorkloadToken string             `json:"workloadToken"`
	Request       PluginRequest      `json:"request"`
	Capabilities  CapabilityEnvelope `json:"capabilities"`
}

type EntryPoint struct {
	Name                string         `json:"name"`
	Type                string         `json:"type"`
	ResourceType        string         `json:"resourceType"`
	ConfigurationSchema map[string]any `json:"configurationSchema"`
	OutputSchema        map[string]any `json:"outputSchema,omitempty"`
}

type DiscoveryResult struct {
	SessionID     string       `json:"sessionId"`
	WorkloadToken string       `json:"workloadToken"`
	EntryPoints   []EntryPoint `json:"entryPoints"`
}

type PluginResponse struct {
	ProtocolVersion string                   `json:"protocolVersion"`
	InvocationID    string                   `json:"invocationId"`
	Output          map[string]any           `json:"output"`
	Logs            []map[string]any         `json:"logs"`
	Errors          []map[string]any         `json:"errors"`
	Checkpoint      map[string]any           `json:"checkpoint,omitempty"`
}

type PluginHandler func(PluginRequest, CapabilityEnvelope) (PluginResponse, error)
